[2단계 - 장바구니 협업] 병민(윤병인) 미션 제출합니다.

airman5573 opened the pull request:
안녕하세요 리트님!  
아직 1단계 코멘트도 해결 못했는데 이렇게 2단계 리뷰 요청을 드리게 되었네요 ㅠㅠ 
`LocalStorage.getItem("accessToken");`에 대한 추상화([comment](https://github.com/woowacourse/react-shopping-cart-prod/pull/18#discussion_r889517772))에 대한 부분과 Layout Template 개선은 아직 해결을 못했습니다. 이부분은 제외하고 봐주시면 감사하겠습니다 :D   

## 백엔드와의 협업
[API명세(상세) - Notion](https://www.notion.so/API-e84292d1f738480bbb7d901f03a51ecd)  
[API명세 - GoogleSheet](https://docs.google.com/spreadsheets/d/1v-4dP1cLWq506oUpKLwsCKQcycxOTKQMczFtZfJ00rQ/edit?usp=sharing)  

백엔드와 협업하면서 많이 즐거웠습니다. 서로 근거를 가지고 주장하면서 설득하고 설득당하는 과정이 즐거웠습니다.  
저희는 회의를 정말 길게 했는데 그래서인지 API명세를 꼼꼼하게 작성했고 서로 그것을 잘 지켜서  
백엔드에서 만들어준 Real API를 연결했을때 큰 문제 없이 잘 작동했습니다.  
나아가 같이 협업하는 소프트 스킬도 조금은 향상 된듯 싶습니다. Level3에도 이 경험을 살려서 잘 해보고 싶다는 생각이 들었습니다.  

## 고민
### Redux...어떻게 사용해야 하는가...
이번 미션에서 거의 모든 상태를 redux store에 넣어 놓고 관리했습니다. 그래서 다음과 같은 장단점을 느꼈습니다.
#### 장점
1. action dispatch -> reducer -> state update 와 같은 순서가 정해져 있어서 디버깅시 적합한 파일에 빠르게 찾아갈 수 있었습니다.
2. 상태를 공유하기 때문에 props drilling이나 context api를 사용할 필요가 없었습니다.

#### 단점
1. 코드가 너무 길어져서 어지러웠습니다. Action작성하고 Reducer작성하다보면 약간 질려버리는 느낌도 들었습니다. Redux-Took-Kit을 빨리 써보고 싶습니다.
2. 상태를 초기화 해야하는 번거로움이 있을것으로 예상됩니다. A페이지에서 비동기 처리후 isSuccess의 상태가 true가 되고, 다른 페이지로 넘어갔을때 isSuccess를 초기화 해주지 않으면, 다시 A페이지로 돌아왔을때 뭔가 해야하는데 isSuccess가 true여서 다른 로직이 실행됩니다. 이거는 정말 어쩔 수 없이 초기화 해줘야 하는것 같은데,,,뭔가 이런식으로 하다보면 코드가 점점 복잡해 지지 않을까? 하는 생각도 듭니다. 예를들어 `페이지를 벗어날때`라는 이벤트가 발생했을때 `초기화`를 해줘야 하는데, 어디서 `초기화`를 했는지  찾기가 좀 어려워지지 않을까...하는 생각이 듭니다. **보통 현업에서 Redux를 쓸때 이 부분이 큰 문제가 되는지 궁금합니다!** + (Redux + Thunks) 대신 다른 괜찮은 조합이 있다면 추천 부탁드립니다! 리덕스 힘드네요..

pocojang committed:
feat: add contents

pocojang committed:
fix typos

pocojang committed:
feat: remove requirements

airman5573 committed:
chore: git 커밋 메세지의 규칙을 설정한다

airman5573 committed:
chore: 프로젝트 기본 설정

1. webpack
2. babel
3. scss
4.emotion
5. eslint & prettier

airman5573 committed:
feat(component): header를 생성한다

airman5573 committed:
feat(ui): home page 생성

airman5573 committed:
feat(cart): 장바구니에 상품 저장하는 기능 추가

airman5573 committed:
feat(redux): redux로 장바구니 상태 관리를 한다

airman5573 committed:
style(responsive): 반응형 디자인을 적용한다

airman5573 committed:
style(component): header를 sticky하게 만든다

airman5573 committed:
fix: link 수정

airman5573 committed:
refactor(style): 불필요한 style삭제

airman5573 committed:
docs(requirements): 기능 요구사항 목록을 생성한다

airman5573 committed:
chore(gitignore): tracking 하지 않을 목록을 추가한다

airman5573 committed:
refactor(redux): action type을 그룹핑한다

airman5573 committed:
refactor(a11y): image에alt속성을 추가한다

airman5573 committed:
refactor(style): css-in-js대신 css-module을 사용한다

airman5573 committed:
feat: live server대신 msw mocking server를 이용한다

airman5573 committed:
feat(ui): 이미지를 로딩하는 동안 indicator를 보여준다

airman5573 committed:
chore(webpack): 웹팩 설정 파일을 분리한다

airman5573 committed:
feat(style): windicss를 적용한다

airman5573 committed:
refactor(environment): 환경 변수를 수정한다

airman5573 committed:
refactor(cleanup): 불필요한 파일을 삭제한다

airman5573 committed:
style(favicon): favicon을 업데이트 한다

airman5573 committed:
refactor(cleanup): 불필요한 파일을 삭제한다

airman5573 committed:
refactor: className을 일관성 있게 작성한다

airman5573 committed:
refactor(cleanup): 절대 경로를 사용한다

airman5573 committed:
refactor(cleanup): 불필요한 className을 삭제한다

airman5573 committed:
refactor(cleanup): 사용하지 않는 파일을 삭제한다

airman5573 committed:
fix(title): 타이틀을 변경한다

youngbeomrhee committed:
Merge pull request #97 from airman5573/airman5573

[1단계 - 장바구니] 병민(윤병인) 미션 제출합니다.

airman5573 committed:
feat(storybook): storybook을 추가한다

airman5573 committed:
refactor(style): windicss 대신 tailwind를 사용한다

airman5573 committed:
refactor(cleanup): 불필요한 코드를 삭제한다

airman5573 committed:
refactor(component): dynamic한 css속성을 부여한다

airman5573 committed:
style(component): 왼쪽 여백을 추가한다

airman5573 committed:
refactor(component): handler함수를 등록한다

airman5573 committed:
refactor(lint): lint 수정

airman5573 committed:
refactor(cleanup): 불필요한 파일을 삭제한다

airman5573 committed:
fix(component): 함수가 있는 경우에만 호출한다

airman5573 committed:
refactor(component): icon에 button을 입힌다

airman5573 committed:
style(component): 기본 크기를 지정한다

airman5573 committed:
style(component): 스타일을 수정한다

airman5573 committed:
refactor(component): input에 variation을 추가한다

airman5573 committed:
feat(component): cart item을 추가한다

airman5573 committed:
refactor(lint): lint를 수정한다

airman5573 committed:
refactor(component): 속성값을 수정한다

airman5573 committed:
chore(storybook): 버전 업데이트를 한다

airman5573 committed:
chore(tailwind): pixel값을 사용한다

tailwind에서 기본으로 제공하는 util대신 pixel단위의 util class를 사용한다

airman5573 committed:
feat(storybook): tailwind를 스토리북에서도 사용한다

airman5573 committed:
style(component): css를 개선한다

airman5573 committed:
refactor(component): default props를 부여한다

airman5573 committed:
style(component): minor change

airman5573 committed:
refactor(component): 외부로부터 className을 받는다

airman5573 committed:
refactor(cleanup): 사용하지 않는 코드를 삭제한다

modal을 삭제한다. 지금당장 사용할 컴포넌트가 아니기 때문이다

airman5573 committed:
refactor(cleanup): 사용하지 않는 코드를 삭제한다

modal을 삭제한다

airman5573 committed:
refactor(component): 일관된 값을 넣어준다

airman5573 committed:
refactor(redux): 초기값을 변경한다

airman5573 committed:
refactor(webpack): 절대 경로를 추가한다

airman5573 committed:
refactor(component): props를 추가한다

체크되었는지 상태를 받는다

airman5573 committed:
refactor(cleanup): 사용하지 않는 파일을 삭제한다

airman5573 committed:
feat(component): 장바구니 form을 추가한다

airman5573 committed:
fix(webpack): 절대 경로를 추가한다

storyboard 뿐만 아니라 webpack에도 추가를 해준다

airman5573 committed:
refactor(typo): minor change

airman5573 committed:
refactor(cleanup): 사용하지 않는 prop을 삭제한다

airman5573 committed:
refactor(naming): 짧은 이름으로 변경한다

이름에 중복이 있기 때문이다

airman5573 committed:
refactor(api): api 구조를 변경한다

1.  title -> name : 상품의 "이름" 이 더 적절해보인다
2. price 단위 제거 : 더 유연하게 사용 가능하고 계산시에 편리하다
3. id -> sku : 상품을
식별하는 요소는 sku이다 id는 DB에서 사용하는 용어이다

airman5573 committed:
chore: 절대 경로를 추가한다

airman5573 committed:
style(component): 이미지의 최대 너비를 제한한다

airman5573 committed:
style(component): highlighter

스타일을 변경한다

ISSUES CLOSED: 기본 색상과 투명도를 수정했다

airman5573 committed:
refactor(utils): 유틸 함수로 분리한다

CartTotal컴포넌트에서 사용될것이기 때문에 분리한다

airman5573 committed:
style(component): 색상을 수정한다

airman5573 committed:
feat(component): 컴포넌트를 추가한다

총 주문 금액과 주문버튼이 있는 CartTotal컴포넌트를 추가한다

airman5573 committed:
fix(component): module을 가져오는 출처를 변경한다

airman5573 committed:
refactor(utils): 유틸 함수로 분리한다

airman5573 committed:
feat(component): 컴포넌트를 추가한다

alert 컴포넌트를 추가했다

airman5573 committed:
refactor(lint): lint를 바탕으로 수정한다

airman5573 committed:
fix(component): 올바른 컴포넌트를 사용한다

anchor대신 react-dom-router의 Link를 사용한다

airman5573 committed:
refactor(component): 상태 관리를 상위에 맡긴다

airman5573 committed:
refactor(component): 상태값이 있는 경우에 렌더링 한다

airman5573 committed:
refactor(component): 컴포넌트를 분리한다

airman5573 committed:
fix: 올바른 속성을 사용한다

airman5573 committed:
fix: 올바른 값을 입력한다

airman5573 committed:
fix(ui): 우선순위를 부여한다

이 부분은 추후에 다시 살펴봐야 한다. 이상하게 일반 태그의 우선순위가 더 높게 책정된다.

airman5573 committed:
chore(tailwind): 추가 css를 설정한다

airman5573 committed:
refactor(component): 외부에서 스타일링을 할 수 있도록 열어놓는다

airman5573 committed:
fix(redux): 장바구니에서 상품 삭제시 localStorage에도 반영한다

airman5573 committed:
refactor(redux): 상위에서 state를 가져온다

주소창에서 바로 장바구니로 넘어간 경우 ProductList를 거쳐가지 않았기 때문에 상품 정보가 없다. 그래서 상위에서 상품 정보를 가져온다

airman5573 committed:
perf(redux): re-render를 막는다

사실 자신이 없다. 정말 re-render를 막아줄까

airman5573 committed:
refactor(cleanup): 사용하지 않는 메뉴를 삭제한다

airman5573 committed:
test(redux): reducer 테스트를 추가한다

airman5573 committed:
Merge branch 'airman5573' of https://github.com/airman5573/react-shopping-cart into airman5573
Merge

airman5573 committed:
docs: 기능 요구 사항을 추가한다

airman5573 committed:
style(responsive): 헤더를 반응형으로 만든다

airman5573 committed:
style(component): 높이를 조정한다

flexbox를 쓰면 쫘악 위아래로 붙는데 이거를 해결한다

airman5573 committed:
style(component): 스크롤이 길어졌을때도 총금액 박스는 붙어있도록 한다

airman5573 committed:
refactor(style): style을 scss에 분리한다

airman5573 committed:
style(responsive): cart page를 반응형으로 개선한다

airman5573 committed:
style: 기본 좌우 padding을 준다

airman5573 committed:
refactor(component): 단독으로 사용도 가능하고 외부와 연결도 되게끔 개선한다

항상 외부에만 연결되어 사용하는 것이 아니라, 단독으로 사용해도 되게끔 한다. 직접 element에 접근해서 값을 가져다가 쓸 수도 있기 때문이다.

airman5573 committed:
chore(storybook): 배포 process를 추가한다

airman5573 committed:
refactor: es6 방식으로 import한다

airman5573 committed:
chore(git): tracking하지 않을 파일을 추가한다

airman5573 committed:
refactor(cleanup): 사용하지 않는 컴포넌트를 삭제한다

airman5573 committed:
refactor(readability): hook을 코드 상단으로 올린다

airman5573 committed:
refactor(component): 컴포넌트를 합친다

가독성을 위해 컴포넌트를 나누지 않고 합친다

airman5573 committed:
style(ui): 이미지 비율을 가로길이에 맞춘다

airman5573 committed:
style(ui): 장바구니 아이콘 배열을 조정한다

airman5573 committed:
fix(component): props를 올바르게 정의한다

airman5573 committed:
perf(component): 불필요한 동일성 체크를 제거한다

airman5573 committed:
perf(redux): 동일성 체크 함수를 개선한다

airman5573 committed:
chore(lint): 설정을 변경한다

dot-notation보다 [key]가 더 자연스러운 경우가 있기 때문에, off시켜준다

airman5573 committed:
test(reducer): test case를 추가한다

airman5573 committed:
chore(babel): version upgrade

airman5573 committed:
refactor(cleanup): 사용하지 않는 파일을 삭제한다

airman5573 committed:
Merge branch 'airman5573' of https://github.com/airman5573/react-shopping-cart-prod into airman5573

airman5573 committed:
refactor(cleanup): 사용하지 않는 module을 삭제한다

airman5573 committed:
chore(lint): lint rule을 합친다

react를 import하지 않아도 jsx를 사용할 수 있게 끔하는 rule을 extend로 합친다.

airman5573 committed:
refactor(cleanup): 사용하지 않는 파일을 삭제한다

airman5573 committed:
refactor(component): react import없이 jsx를 사용한다

airman5573 committed:
refactor(lint): lint를 적용한다

airman5573 committed:
refactor(component): react를 import하지 않고 jsx를 사용한다

airman5573 committed:
chore(lint): lint rule을 변경한다

airman5573 committed:
refactor(component): 컴포넌트를 합친다

Menu컴포넌트를 굳이 분리 할 필요가 없어서 합쳤다

airman5573 committed:
refactor(cleanup): 사용하지 않는 파일을 삭제한다

airman5573 committed:
refactor(component): 버튼 컴포넌트를 개선한다

1. type을 동적으로 입력 받을 수 있다
2. size 속성을 추가한다

airman5573 committed:
feat(component): input 컴포넌트를 추가한다

airman5573 committed:
feat(component): labeledInput 컴포넌트를 추가한다

airman5573 committed:
feat(hooks): useForm 훅을 추가한다

airman5573 committed:
feat(template): authFormTemplate을 추가한다

airman5573 committed:
style: 전역 font를 변경한다

airman5573 committed:
feat(page): 로그인 페이지를 추가한다

airman5573 committed:
feat(page): 회원가입 페이지를 추가한다

airman5573 committed:
refactor(component): 버튼에 속성을 추가한다

airman5573 committed:
refactor(component): divider에 속성을 추가한다

airman5573 committed:
refactor(component): 컴포넌트를 삭제한다

범용성 있는 FontawesomeIconButton으로 대체한다

airman5573 committed:
refactor(page): ui 개선

airman5573 committed:
fix(component): 존재하는 컴포넌트를 사용한다

airman5573 committed:
feat(page): 마이페이지를 추가한다

airman5573 committed:
refactor(hook): useForm을 개선한다

1. submit시 최근값을 돌려준다.
2.pattern validation을 지원한다

airman5573 committed:
feat: localStorage class를 구현한다

airman5573 committed:
feat(remote): remote api를 추가한다

각 페이지에서 api 요청을 보낸다

airman5573 committed:
feat(component): login 상태에 따라 header menu를 다르게 보여준다

airman5573 committed:
refactor(remote): 앱이 로드될때 로그인 상태를 업데이트한다

airman5573 committed:
feat(component): 로그인 상태에 따라 페이지 접근을 제한하는 컴포넌트를 추가한다

airman5573 committed:
refactor(cleanup): 사용하지 않는 코드를 삭제한다

airman5573 committed:
chore(lint): rule을 추가한다

alert를 허용한다

airman5573 committed:
refactor(msw): 잘못된 요청에 대응한다

airman5573 committed:
docs: 기능 요구사항 추가

airman5573 committed:
refactor: 기본값을 삭제한다

airman5573 committed:
refactor(lint): lint suggestion 해결

airman5573 committed:
refactor(cleanup): 지원하지 않는 기능을 삭제한다

airman5573 committed:
fix(component): 에러 처리를 하지 않는다

airman5573 committed:
refactor(component): button에 속성을 추가한다

disabled 속성을 추가한다

airman5573 committed:
feat(component): input 입력값에 대한 feedback을 추가한다

airman5573 committed:
fix(ui): 컴포넌트의 사이즈를 조정한다

airman5573 committed:
fix(msw): 기본 값을 넣어준다

airman5573 committed:
feat(hook): useForm hook에 custom validator 기능을 추가한다

airman5573 committed:
refactor(page): myPage의 validation을 개선한다

1. input사이의 간격을 조절한다
2. 입력값에 대한 유효성 검사 후 feedback을 보여준다

airman5573 committed:
refactor(hook): useForm에서 input의 defaultValue를 지원한다

airman5573 committed:
refactor(page): form component를 분리한다

airman5573 committed:
refactor(cleanup): 중복을 제거한다

이미 {...productObjs[id]}에서 name, thunbmail_image, price를 넘겨주고 있기 때문에
추가적으로 또 넘겨줄 필요는 없다.

airman5573 committed:
docs: 기능 구현 목록을 추가한다

airman5573 committed:
refactor: product schema를 변경한다

sku대신 id를 사용한다

airman5573 committed:
refactor(component): disabled 속성을 사용하지 않는다

airman5573 committed:
docs: 기능 구현 목록을 수정한다

airman5573 committed:
feat(msw): handler를 추가한다

장바구니, 상품 단건 조회, 주문 목록 관련 api를 추가한다

airman5573 committed:
refactor(schema): product schema를 변경한다

thumbnail_image를 thumbnailImage로 변경한다

airman5573 committed:
refactor(redux): product관련 reducer를 분리한다

airman5573 committed:
refactor(lint): 파일에서 lint rule 비활성화 한다

airman5573 committed:
refactor(redux): reducer를 분리한다

airman5573 committed:
chore(webpack): 설정을 변경한다

publicPath를 지정해서 detail 페이지에서도 root에 있는 파일을 요청하도록 한다.

airman5573 committed:
refactor(msw): error처리를 개선한다

airman5573 committed:
fix: mocking data를 정리한다

airman5573 committed:
refactor(redux): thunks를 활용한다

airman5573 committed:
refactor(naming): 함수명을 변경한다

productId가 아닌 cartItemId을 사용한다

airman5573 committed:
feat(component): 상세페이지를 추가한다

airman5573 committed:
chore(lint): rule을 추가한다

airman5573 committed:
refactor(redux): 기본 초기 상태를 추가한다

airman5573 committed:
fix(redux): cartItem의 기본 상태를 추가한다

airman5573 committed:
fix(msw): 재고 상태를 추가한다

airman5573 committed:
feat(redux): 주문 관련 reducer를 추가한다

airman5573 committed:
refactor(component): authGuard를 개선한다

로그인 상태에 따라 페이지에 접근을 제한한다

airman5573 committed:
feat(page): 주문 목록 페이지를 생성한다

airman5573 committed:
build: env파일 수정

airman5573 committed:
refactor: local api를 사용한다

airman5573 committed:
fix(component): 올바른 통화를 사용한다

airman5573 committed:
feat(msw): 주문 성공시에 카트에서 상품을 삭제한다

airman5573 committed:
refactor(redux): 성공적으로 이름변경시 사용자에게 알려준다

airman5573 committed:
fix(webpack): 절대경로를 추가한다

airman5573 committed:
refactor(readability): 개행을 추가한다

airman5573 committed:
refactor(cleanup): 불필요한 주석을 제거한다

airman5573 committed:
refactor(cleanup): 불필요한 console을 제거한다

airman5573 committed:
refactor(readability): 개행을 추가한다

airman5573 committed:
Co-authored-by: Soyi Jeon <soyi.dev@gmail.com>

airman5573 committed:
refactor(cleanup): 사용하지 않는 파일을 삭제한다

lsw1164 commented on src/utils/fetcher.js:
```diff
@@ -0,0 +1,65 @@
+
+class Fetcher {
+  // eslint-disable-next-line no-undef
+  static BASE_URL = API_URL;
```
import가 없던데 API_URL는 어떤거죠?

airman5573 replied:
[.env.dev](https://github.com/airman5573/react-shopping-cart-prod/blob/airman5573/env/.env.dev) 파일에 있습니다! 

그리고 [webpack에서 DefinePlugin을 사용해](https://github.com/airman5573/react-shopping-cart-prod/blob/859238ae576231754bc2115110e4c1f91c35bdab/webpack/webpack.dev.js#L16) .env.dev파일에 있는 API_URL을 main.js파일에 넣었습니다.

lsw1164 replied:
오... webpack 기능이군요 DefinePlugin은 처음봐서 그런지 블랙박스 같은 느낌이네요.
IDE에서 쉽게 확인 가능한가요?

airman5573 replied:
IDE에서 전체파일 문자열 검색으로만 확인 가능합니다 ㅠ  

그리고 제가 조금 잘못 말씀드린것이, API_URL을 main.js파일에 넣는다기 보다는 대체 한다고 합니다!  

```javascript
if (!PRODUCTION) {
  console.log('Debug info');
}

if (PRODUCTION) {
  console.log('Production log');
}

// build후

if (!true) {
  console.log('Debug info');
}
if (true) {
  console.log('Production log');
}
```

lsw1164 commented on src/utils/fetcher.js:
```diff
@@ -0,0 +1,65 @@
+  // eslint-disable-next-line no-undef
+  static BASE_URL = API_URL;
+
+  static accessToken = LocalStorage.getItem("accessToken");
+
+  static updateAccessToken(accessToken) {
+    this.accessToken = accessToken;
+  }
+
+  static getHeadersWithAccessToken = () => {
+    const headers = this.accessToken
+      ? {
+          Authorization: `Bearer ${this.accessToken}`,
+        }
+      : {};
+    return headers;
+  };
+
```
두 가지 의문이 들었는데요.

1. Fetcher가 accessToken의 관리 책임까지 갖고 있는 걸까요?

2. Fetcher가 LocalStorage에 의존성을 갖고 있는데요.
updateAccessToken에서 accessToken을 LocalStoreage에 저장하지 않는 이유가 있을까요?

airman5573 replied:
1. **Fetcher가 accessToken의 관리 책임까지 갖고 있는 걸까요?**

조금 애매합니다 ㅠ  
Fetcher를 쓸때 accessToken이 매우 자주 사용되기 때문에, LocalStorage에서 accessToken을 가져와서 내부적으로 캐싱하고 때에따라 캐싱된 값을 업데이트 해주고 있습니다. 그래서 Fetcher가 accessToken을 "관리"한다고 해야할지 조금 의문이 듭니다. accessToken을 관리 하는건 LocalStorage이고 Fetcher는 I/O비용을 줄이기 위해 캐싱해서 사용하는 정도가 아닌가 싶습니다.

2. **Fetcher가 LocalStorage에 의존성을 갖고 있는데요.
updateAccessToken에서 accessToken을 LocalStoreage에 저장하지 않는 이유가 있을까요?**

Fetcher가 accessToken을 관리하는 주체가 아니기 때문입니다.  


리뷰어님의 질문을 듣고, Fetcher와 LocalStorage의 의존성을 떨어뜨려야 겠다는 생각이 들었습니다. 또한, 현재 Fetcher를 범용적으로 쓰기가 어렵다는 생각도 들었습니다.  

아래와같이 static을 제거하고 필요할때 initialize해서 쓰는편이 더 나을까요?
```javascript
class Fetcher {
  constructor(accessToken, baseUrl = API_URL) {
    this.baseUrl = baseUrl;
    this.accessToken = accessToken;
  }

  getHeadersWithAccessToken = () => {
    const headers = this.accessToken
      ? {
          Authorization: `Bearer ${this.accessToken}`,
        }
      : {};
    return headers;
  };

  get = (endpoint) => {};

  post = (endpoint, body) => {};

  patch = (endpoint, body) => {};

  delete = (endpoint, body) => {};
}
```

lsw1164 replied:
그렇게 하면 의존성은 해결하겠군요 👍
토큰을 외부에서 주입하게 되면 instance을 계속 새로 만들어야 될 거 같은데. 
그냥 get, post, patch, delete 각 메서드에 토큰 파라미터 받아서 utils 함수로 만들어도 될 거 같다는 생각이 들었습니다.

airman5573 replied:
말씀하신대로 accessToken을 각 메서드에 파라미터로 받아 사용했습니다.

```javascript
static get = ({ endpoint, accessToken }) => {
  return fetch(`${this.BASE_URL}/${endpoint}`, {
    headers: { ...this.getHeadersWithAccessToken(accessToken) },
  });
};
```

다만, `LocalStorage.getItem("accessToken")`의 반복이 심해서
```javascript
const accessToken = LocalStorage.getItem("accessToken");
const response = await Fetcher.get({ endpoint: "mycarts", accessToken });
```

[AccessTokenStorage](https://github.com/airman5573/react-shopping-cart-prod/blob/9bd80122bfdba29a5555ad18dd7879b0b8bbc8a5/src/storage/accessTokenStorage.js#L1)를 만들었습니다.  
[AccessTokenStorage안에서 LocalStorage를 사용했었는데](https://github.com/woowacourse/react-shopping-cart-prod/pull/60/commits/f4149008875ce979240494536dceffa91f7d87c3#diff-92f87a175741fed3774811b484ba9bb467484817b6154ef1d4a7a51409af72ec), 이렇게 되면 또 의존성이 발생해서, [외부에서 storage를 주입받는 식으로 개선](https://github.com/woowacourse/react-shopping-cart-prod/pull/60/commits/9bd80122bfdba29a5555ad18dd7879b0b8bbc8a5)했습니다.

lsw1164 commented on src/redux/utils/ApiError.js:
```diff
@@ -0,0 +1,15 @@
+class ApiError extends Error {
```
Api 예외 상황에서 쓰는 Error 객체군요 👍

lsw1164 commented on src/index.jsx:
```diff
@@ -18,39 +18,11 @@ function prepareMSW() {
-  const store = await setupStore();
+  const store = createStore(reducer, applyMiddleware(thunk));
+
+  const accessToken = LocalStorage.getItem("accessToken");
+  accessToken && (await getUser()(store.dispatch));
```
getUser를 기다릴 때까지 사이트가 아무것도 그리지 않는건가요? 🤔

airman5573 replied:
네, 현재는 유저정보를 갖고 온 뒤에 App을 렌더링 하도록 설계했습니다.  
왜냐하면 처음에 잠깐만 기다리면 이후에는 로그인 유무를 검사하는 로직이 없어 깜빡거리는 현상이 줄어들것이라 생각했기 때문입니다.  (특히 header)  
하지만 리뷰어님의 질문을 듣고 '이게 최선일까?' 하는 생각이 들었습니다.  
그래서 이 문제를 해결하는 여러가지 방법을 조사해보았습니다.

## 깜빡여도 괜찮다
**라프텔**  
![laftel-loading-gif](https://develoger.kr/wp-content/uploads/loding-laftel.gif)

**New York Times**
![new-york-times-gif](https://develoger.kr/wp-content/uploads/new-york-times.gif)

라프텔과 NYT모두 로딩 후에 로그인된 유저인지를 확인하는듯 싶습니다.  
라프텔은 괜찮은데, NYT는 로그인 버튼이 꽤나 오래 표시되어 있어서 UI/UX적으로 개선이 필요하다고 생각했습니다.

## 레이아웃으로 승부하기
**Airbnb**
![airbnb](https://develoger.kr/wp-content/uploads/airbnb.gif)

보시다싶이 데스크탑 모드에서도 햄버거 메뉴를 사용하기 때문에 레이아웃에 변화가 없습니다. 이미지만 비동기적으로 바뀔 뿐입니다. 내부적으로는 페이지 로딩 후에 추가로 로그인 상태인지 확인하고 re-render되는것 같습니다만, 사용자에게 그것이 표출되지는 않습니다. 

## 안깜빡이기
**velog**
![velog](https://develoger.kr/wp-content/uploads/velog.gif)

velog는 깜빡이지 않습니다. 레이아웃도 바뀌지 않습니다. chrome에서 js를 꺼도 "로그인" 버튼이 아닌 "새 글 작성" 버튼이 나왔습니다. 저는 어떻게 이럴 수 있는지 궁금했습니다. 다행히! velog는 [오픈 소스](https://github.com/velopert/velog-client)여서 코드를 까볼 수 있었습니다.

### [header.jsx](https://github.com/velopert/velog-client/blob/8f4a030aa69d30674e96584d6b5e64ab6e0b4889/src/components/base/Header.tsx#L20)
3항연산자로 로그인 안된 상태면 "로그인" 버튼을, 로그인 된 상태면 "새 글 작성" 버튼을 보여줍니다.

### [ssrMiddleware.js](https://github.com/velopert/velog-client/blob/8f4a030aa69d30674e96584d6b5e64ab6e0b4889/src/server/ssrMiddleware.ts#L4)

```javascript
const result = await serverRender({
  url: ctx.url,
  cookie: ctx.headers.cookie,
  loggedIn: !!(
    ctx.cookies.get('refresh_token') || ctx.cookies.get('access_token')
  ),
});
```
ssrMiddleware에서 serverRender를 호출합니다.

### [serverRender.js](https://github.com/velopert/velog-client/blob/8f4a030aa69d30674e96584d6b5e64ab6e0b4889/src/server/serverRender.tsx#L86)

serverRender안에서는 서버 사이드 렌더링을 합니다. 

```javascript
const Root = (
    <ChunkExtractorManager extractor={extractor}>
      <HelmetProvider context={helmetContext}>
        <StyleSheetManager sheet={sheet.instance}>
          <Provider store={store}>
            <UncachedApolloProvider client={client}>
              <ApolloProvider client={client}>
                <StaticRouter location={url} context={context}>
                  <App />
                </StaticRouter>
              </ApolloProvider>
            </UncachedApolloProvider>
          </Provider>
        </StyleSheetManager>
      </HelmetProvider>
    </ChunkExtractorManager>
  );
```

여기서 [getDataFromTree](https://www.apollographql.com/docs/react/api/react/ssr/#getdatafromtree)로 Root에서 apollo client로 호출된 요청들을 전부 처리해줍니다.
```javascript
try {
    await getDataFromTree(Root);
  } catch
```

### [useUserLoader.jsx](https://github.com/velopert/velog-client/blob/8f4a030aa69d30674e96584d6b5e64ab6e0b4889/src/containers/base/hooks/useUserLoader.tsx#L11)

여기서 user정보를 가져옵니다

```javascript
const getCurrentUser = useQuery<{ auth: CurrentUser }>(GET_CURRENT_USER);
  const prevUser = useSelector((state: RootState) => state.core.user);

  const user = getCurrentUser.data ? getCurrentUser.data.auth : undefined;

  if (ssrEnabled && user) {
    dispatch(setUser(user));
  }
```

즉, 서버에서 cookie를 바탕으로 유저를 가져와서 렌더링까지 해준 html string을 돌려줬기 때문에 깜빡임 없이 "새 글 작성" 버튼이 나왔다고 생각합니다.

## 결론
1. index.jsx에서 사용자 정보를 가져오지 않고, 사용자 정보가 필요한 곳에서 요청을 보내겠습니다.(수정 커밋 : [8003384](https://github.com/woowacourse/react-shopping-cart-prod/pull/60/commits/800338409edfadcf1e7abbd32bdc58afa178272a))
2. SSR이 제일 좋은 선택지인것 같은데, 지금 미션에서는 구현하기 어렵기 때문에 잠깐 data fetching하는 동안 [header 메뉴에서 아무것도 안보이게 하고, 이후에 가져오면 메뉴를 보이도록](https://github.com/airman5573/react-shopping-cart-prod/blob/800338409edfadcf1e7abbd32bdc58afa178272a/src/components/header/Header.jsx#L68) 하는것이 최선이라고 생각했습니다. 

혹시 더 괜찮은 방법이 있는지, 리뷰어님의 의견이 궁금합니다!

lsw1164 replied:
여러 래퍼런스를 참고 및 분석해서 현재 상황에서 적절한 방법을 찾으신 거 같아요 👍

lsw1164 reviewed (changes requested):
병민님, 코드 잘봤습니다 
redux 사용법도 익숙해지신 거 같고 장점을 잘 파악하고 계시네요👍
코멘트 확인해주세요!

++
> 상태를 초기화 해야하는 번거로움이 있을것으로 예상됩니다. A페이지에서 비동기 처리후 isSuccess의 상태가 true가 되고, 다른 페이지로 넘어갔을때 isSuccess를 초기화 해주지 않으면, 다시 A페이지로 돌아왔을때 뭔가 해야하는데 isSuccess가 true여서 다른 로직이 실행됩니다. 이거는 정말 어쩔 수 없이 초기화 해줘야 하는것 같은데,,,뭔가 이런식으로 하다보면 코드가 점점 복잡해 지지 않을까? 하는 생각도 듭니다. 예를들어 페이지를 벗어날때라는 이벤트가 발생했을때 초기화를 해줘야 하는데, 어디서 초기화를 했는지 찾기가 좀 어려워지지 않을까...하는 생각이 듭니다. 보통 현업에서 Redux를 쓸때 이 부분이 큰 문제가 되는지 궁금합니다! + (Redux + Thunks) 대신 다른 괜찮은 조합이 있다면 추천 부탁드립니다! 리덕스 힘드네요..

페이지 벗어날 때 초기화 하고 싶다면 [useEffect cleanup](https://ko.reactjs.org/docs/hooks-effect.html#example-using-hooks-1)을 활용해서 초기화 로직을 실행하면 될 거 같아요.



어디서 초기화 했는지 찾기 어려운 건 어떤 의미일까요? 
코드 검색이 불편해서 그럴까요?

비동기 호출 관련해서 Thunks 대신 도움 받을 수 있는 라이브러리는 react-query 나중에 공부해 보시면 어떨까요? ㅎㅎ

airman5573 committed:
refactor: 코드 품질을 개선한다

불필요한 조건문을 줄인다. validation객체의 필드들에 기본적으로 undefined라도 있기 때문에 구조분해 할당으로
코드를 더 짧게 쓸 수 있다.

airman5573 committed:
refactor(readability): 개행을 추가한다

airman5573 committed:
refactor(ui): 사용자 정보를 필요한 곳에서 fetch한다

lsw1164 reviewed (changes requested):
-

airman5573 committed:
refactor: 의존성을 분리한다

Fetcher와 LocalStorage간의 의존성을 분리한다

airman5573 committed:
refactor(storage): layer를 추가한다

access token만 관리하는 layer를 추가한다

airman5573 committed:
refactor: 의존성을 낮춘다

AccessTokenStorage와 LocalStorage간의 의존성을 낮춘다

airman5573 commented:
> 어디서 초기화 했는지 찾기 어려운 건 어떤 의미일까요?
> 코드 검색이 불편해서 그럴까요?

제가 너무 막연한 질문을 드렸었네요!  

페이지를 벗어날때, 혹은 특정 이벤트가 발생했을때 상태를 초기화 해줘야 한다고 생각했습니다.  
다시말해서, useEffect의 cleanup함수 혹은 이벤트 핸들러에서 상태를 초기화 하는 action이 dispatch되야 할텐데,  
이렇게 되면 나중에 코드를 봤을때 어디서 상태를 초기화 하는지 찾기 어렵지 않을까...하는 걱정이 있었습니다. (뭔가 자동으로 되길 바랬던 것이었습니다)  
그런데 이것은 어쩔 수 없는 부분이기도 하고, 그렇게 찾기 어렵지도 않을 것 같다는 생각이 듭니다 :D

lsw1164 reviewed (approved):
병민님, 피드백 반영한 거 확인했습니다.
여러 래퍼런스도 참고하시고 더 좋은 코드에 대해 계속 고민하는 모습을 보니 
앞으로의 성장이 기대되네요 :)
Approve 하겠습니다. 남은 과정도 화이팅입니다~
